docs/debugging.md

```diff
@@ -0,0 +1,169 @@
+
+* Systematic testing
+* Crash logging (A flight recorder that provides the last few seconds of logging)
+
+Both of these can produce high detail logging of the runtime operations.
+
+It is considerably easier to debug a crash in systematic testing, than in standard concurrent execution.
+The majority of tests of the runtime cover a large number of pseudo-random traces, so that the failures can easily be replayed.  The code in
+
+* [src/rt/test/harness.h/SystematicTestingHarness constructor](https://github.com/microsoft/verona/blob/82a00c9d7e0e4f12b08e2b829a599a9ef94c1402/src/rt/test/harness.h#L38-L70)
+* [src/rt/test/harness.h/SystematicTestingHarness::run](https://github.com/microsoft/verona/blob/82a00c9d7e0e4f12b08e2b829a599a9ef94c1402/src/rt/test/harness.h#L73-L114)
+
+highlights all of the features for running with the various logging and random seed approaches.
+
+To compile the run-time with systematic testing you can pass CMake `-DUSE_SYSTEMATIC_TESTING=On`, 
+and then in your program call
+```C++
+Systematic::enable_logging();
+```
+to print the logging, and
+```C++
+Scheduler::get().set_seed(seed);
+```
+specifies the seed for the pseudo-random interleaving.
+The seed to interleaving should be consistent across platforms.
+
+To enable crash logging you can pass CMake `-DUSE_CRASH_LOGGING=On`
+and in your program call
+```C++
+Systematic::enable_crash_logging();
+```
+to enable its output.
+
+The two modes are independent, and both can be enabled if required.
+
+Some build artifacts automatically have these settings enabled by default including some of the runtime unit tests, veronac-sys and interpreter-sys.
+
+The systematic testing log has the earliest entry first, while the crash logging is in reverse, and contains timestamps.
+This is a quirk of how the crash log is reconstructed after the runtime crashes.
+
+Both logs use the same printing in the runtime:
+```C++
+Systematic::cout() << "My messages" << ... << std::endl;
+```
+For, crash log the pretty print functions are not run until the crash, so they
+should not depend on any state other than the `uintptr_t` sized argument they
+are provided.  Here is an example log from systematic testing:
+```
+   3      Scheduled Cown: 0000400001010470 (EPOCH_A)
+  2       Unscheduling Cown: 0000400001010320
+    4     Popped cown:0000400001010400
+    4     Scheduled Cown: 0000400001010400 (EPOCH_A)
+    4     LD protocol loop
+```
+The left hand columns containing `2`, `3`, and `4` represent which thread is executing.
+The rest of the line is the message that is being logged.
+The first line comes from the following line:
+```C++
+        Systematic::cout() << "Scheduled Cown: " << cown << " ("
+                           << cown->get_epoch_mark() << ")" << std::endl;
+```
+It prints the address of the `cown` and the `get_epoch_mark` on that object.
+
+If neither systematic testing or the crash log is enabled, then all the logging
+should be erased by the compiler.
```

Comment:
```suggestion
should be erased by the runtime.
```

I'm guessing you meant "runtime" and not "compiler" here, but feel free to ignore met if I'm wrong